xarray/tests/test_sparse.py

```diff
@@ -0,0 +1,647 @@
+        Coordinates:
+            y        (x) int64 ...
+        Dimensions without coordinates: x""")
+        assert expected == repr(a)
+
+    def test_dataset_repr(self):
+        ds = xr.Dataset(
+            data_vars={'a': ('x', COO.from_numpy(np.ones((4))))},
+            coords={'y': ('x', COO.from_numpy(np.arange(4)))})
+        expected = dedent("""\
+        <xarray.Dataset>
+        Dimensions:  (x: 4)
+        Coordinates:
+            y        (x) int64 ...
+        Dimensions without coordinates: x
+        Data variables:
+            a        (x) float64 ...""")
+        assert expected == repr(ds)
+
+    def test_dataarray_pickle(self):
+        a1 = xr.DataArray(
+            COO.from_numpy(np.ones((4))),
+            dims=['x'],
+            coords={'y': ('x', COO.from_numpy(np.arange(4)))})
+        a2 = pickle.loads(pickle.dumps(a1))
+        assert_identical(a1, a2)
+
+    def test_dataset_pickle(self):
+        ds1 = xr.Dataset(
+            data_vars={'a': ('x', COO.from_numpy(np.ones((4))))},
+            coords={'y': ('x', COO.from_numpy(np.arange(4)))})
+        ds2 = pickle.loads(pickle.dumps(ds1))
+        assert_identical(ds1, ds2)
+
+    @pytest.mark.xfail(reason="No implementation of np.einsum")
+    def test_dot(self):
+        a1 = self.xp_xr.dot(self.xp_xr[0])
+        a2 = self.sp_ar.dot(self.sp_ar[0])
+        assert_equal(a1, a2)
+
+    @pytest.mark.xfail(reason="Groupby reductions produce dense output")
+    def test_groupby(self):
+        x1 = self.ds_xr
+        x2 = self.sp_xr
+        m1 = x1.groupby('x').mean(xr.ALL_DIMS)
+        m2 = x2.groupby('x').mean(xr.ALL_DIMS)
+        assert isinstance(m2.data, sparse.SparseArray)
+        assert np.allclose(m1.data, m2.data.todense())
+
+    @pytest.mark.xfail(reason="Groupby reductions produce dense output")
+    def test_groupby_first(self):
+        x = self.sp_xr.copy()
+        x.coords['ab'] = ('x', ['a', 'a', 'b', 'b'])
+        x.groupby('ab').first()
+        x.groupby('ab').first(skipna=False)
+
+    @pytest.mark.xfail
+    def test_reindex(self):
+        x1 = self.ds_xr
+        x2 = self.sp_xr
+        for kwargs in [{'x': [2, 3, 4]},
+                       {'x': [1, 100, 2, 101, 3]},
+                       {'x': [2.5, 3, 3.5], 'y': [2, 2.5, 3]}]:
+            m1 = x1.reindex(**kwargs)
+            m2 = x2.reindex(**kwargs)
+            assert np.allclose(m1, m2, equal_nan=True)
+
+    @pytest.mark.xfail
+    def test_merge(self):
+        x = self.sp_xr
+        y = xr.merge([x, x.rename('bar')]).to_array()
+        assert isinstance(y, sparse.SparseArray)
+
+    @pytest.mark.xfail
+    def test_where(self):
+        a = np.arange(10)
+        cond = a > 3
+        xr.DataArray(a).where(cond)
+
+        s = COO.from_numpy(a)
+        cond = s > 3
+        xr.DataArray(s).where(cond)
+
+        x = xr.DataArray(s)
+        cond = x > 3
+        x.where(cond)
+
+
+# Do we even want this?
```

Comment:
Right now, we load every "dimension coordinate" into memory as a `pandas.Index`. But for other coordinates, sure why not?